docs/development/design/fsm.md

```diff
@@ -0,0 +1,82 @@
+# Nexd State Machine
+
+> [Issue #885](https://github.com/nexodus-io/nexodus/issues/885)
+
+## Summary
+
+State management in nexd is becoming more complicated.
+We need a mechanism to allow us to simplify that logic and not rely on crash/restart to retrigger things.
+
+## Proposal
+
+I propose that we implement a Finite State Machine inside of nexd that operates as follows:
+
+```mermaid
+---
+title: Nexd State Machine
+---
+stateDiagram-v2
+    [*] --> UNAUTHENTICATED
+    UNAUTHENTICATED --> AUTHENTICATING: "login"
+    AUTHENTICATING --> AUTHENTICATED: "login_success"
+    AUTHENTICATING --> UNAUTHENTICATED: "login_failed"
+    AUTHENTICATING --> UNAUTHENTICATED: "logout"
+    AUTHENTICATED --> REGISTERING: "register_device"
+    AUTHENTICATED --> UNAUTHENTICATED: "logout"
+    REGISTERING --> REGISTERED: "device_registered"
+    REGISTERING --> UNREGISTERED: "registration_failed"
+    REGISTERED --> UNAUTHENTICATED: "logout"
+    UNREGISTERED --> REGISTERING: "register_device"
+    UNREGISTERED --> UNAUTHENTICATED: "logout"
+    REGISTERED --> UP: "up"
+    REGISTERED --> UNAUTHENTICATED: "logout"
+    REGISTERED --> PEERING: "fetch_peers"
+    PEERING --> UNREGISTERED: "down"
```

Comment:
Would going to `UNREGISTERED` imply removing its own device record?

I don't think we'd want to remove it, since we'd want the device to keep its IPs. In that case `UNREGISTERED` doesn't seem quite right.

Maybe it should be: PEERING --> REGISTERED? If the device is administratively down, it would prevent a transition out of REGISTERED.

... but then REGISTERED doesn't seem to capture the state very well, either.  Maybe we need a `DOWN` that is almost identical to REGISTERED, but some transitions are blocked until it's administratively up again?

Reply:
good points. I've added a DOWN state that is like REGISTERED but can only be entered via admin action (i.e stop).
I've stopped anything transitioning to UNREGISTERED, so we'll never unregister the device.
In future we might want to offer the option to do so at some point, but i don't see a use case for it.

The only way out of DOWN is via a "fetch_peers" transition which will eventually get you back into UP - unless anything bad happens along the way.